Meta/Lagom/Tools/LibJSGCVerifier/src/CellsHandler.cpp

```diff
@@ -0,0 +1,149 @@
+/*
+ * Copyright (c) 2023, Matthew Olsson <mattco@serenityos.org>
+ *
+ * SPDX-License-Identifier: BSD-2-Clause
+ */
+
+#include "CellsHandler.h"
+#include <clang/AST/DeclCXX.h>
+#include <clang/AST/Type.h>
+#include <clang/ASTMatchers/ASTMatchFinder.h>
+#include <clang/ASTMatchers/ASTMatchers.h>
+#include <clang/Basic/Specifiers.h>
+#include <clang/Frontend/CompilerInstance.h>
+#include <filesystem>
+#include <llvm/Support/Casting.h>
+#include <llvm/Support/raw_ostream.h>
+
+CollectCellsHandler::CollectCellsHandler()
+{
+    using namespace clang::ast_matchers;
+
+    m_finder.addMatcher(
+        traverse(
+            clang::TK_IgnoreUnlessSpelledInSource,
+            cxxRecordDecl(
+                decl().bind("record-decl"),
+                isDerivedFrom(hasName("::JS::Cell")))),
+        this);
+}
+
+bool CollectCellsHandler::handleBeginSource(clang::CompilerInstance& ci)
+{
+    auto const& source_manager = ci.getSourceManager();
+    ci.getFileManager().getNumUniqueRealFiles();
+    auto file_id = source_manager.getMainFileID();
+    auto const* file_entry = source_manager.getFileEntryForID(file_id);
+    if (!file_entry)
+        return false;
+
+    auto current_filepath = std::filesystem::canonical(file_entry->getName().str());
+    llvm::outs() << "Processing " << current_filepath.string() << "\n";
+
+    return true;
+}
+
+static bool record_inherits_from_cell(clang::CXXRecordDecl const& record)
+{
+    if (!record.isCompleteDefinition())
+        return false;
+
+    bool inherits_from_cell = false;
+    record.forallBases([&](clang::CXXRecordDecl const* base) -> bool {
+        if (base->getQualifiedNameAsString() == "JS::Cell") {
+            inherits_from_cell = true;
+            return false;
+        }
+        return true;
+    });
+    return inherits_from_cell;
+}
+
+static bool validate_gcptr_field(clang::QualType type)
+{
+    if (auto const* elaborated_type = llvm::dyn_cast<clang::ElaboratedType>(type.getTypePtr()))
+        type = elaborated_type->desugar();
+
+    auto const* specialization = type->getAs<clang::TemplateSpecializationType>();
+    if (!specialization)
+        return true;
+
+    auto template_type_name = specialization->getTemplateName().getAsTemplateDecl()->getName();
+    if (template_type_name != "GCPtr" && template_type_name != "NonnullGCPtr")
+        return true;
+
+    if (specialization->getNumArgs() != 1)
+        return true;
+
+    auto const& type_arg = specialization->getArg(0);
+    auto const* record_type = type_arg.getAsType()->getAs<clang::RecordType>();
+    if (!record_type)
+        return true;
+
+    auto const* record_decl = record_type->getAsCXXRecordDecl();
+    if (!record_decl->hasDefinition())
+        return true;
```

Comment:
I think you're right yeah, that case isn't caught. I did a test that was all in the same file and it worked fine, but I guess when you split the file up it no longer works. I'm not totally sure how to fix this at the moment, it doesn't work even if I give clang both file paths to look up. I was under the impression that they'd be resolved at the point `::run` is ran, but apparently not. I'll keep looking into it.

Reply:
@nico do you have any ideas on how to catch this problem using clang's C++ API? I saw your name in the git history for chromium's similar tools (and I think you linked me to them before :) ).

My understanding is that your tool's `run()` will be called when an AST match occurs, and that the CompilerInstance/ASTContext is only extant until `handleEndSource` is called on the tool (which you didn't override :)). 

So currently, the tool only looks at `clang::CXXRecordDecl`s that are themselves `JS::Cell`s, and checks their own members for errant pointers. But if the data members of the classes that you're checking are pointers to incomplete types (as we tend to do .. :) ). Then we won't know about them. I think that only checking `.cpp` files might help, as presumably the classes we're using would need their complete definition in order to allocate them on the JS Heap, but that might not be true.

Reply:
Doing the analysis in the `handleEndSource` callback might also help? As at that point, we can ask the QualType for their canonicalDecl or whatever the correct API is to get the actual definition of each type before checking its inheritance tree.

Reply:
I'm pretty sure the solution here is to remove the optimization where I only visit each class once. This way we'll visit a class every time it's used (i.e. from a cpp file), and it is effectively guaranteed that it will have a complete definition in at least one of them. I tested by adding a `struct Foo;` and `GCPtr<Foo> m_dummy` to `StringIterator.h` and a corresponding `struct Foo {};` to `StringIterator.cpp` and it was detected, so I think we're good here. While I'm at it I'll implement your suggestion to only look at .cpp files, since the headers don't really do anything for us.